cpp/include/raft/spatial/knn/ball_cover.hpp

```diff
@@ -0,0 +1,133 @@
+namespace knn {
+
+template <typename value_idx = int64_t, typename value_t,
+          typename value_int = int>
+inline void rbc_build_index(const raft::handle_t &handle,
+                            BallCoverIndex<value_idx, value_t> &index, int k) {
+  ASSERT(index.n == 2,
+         "Random ball cover currently only works in 2-dimensions");
+  if (index.metric == raft::distance::DistanceType::Haversine) {
+    detail::rbc_build_index(handle, index, k, detail::HaversineFunc());
+  } else if (index.metric == raft::distance::DistanceType::L2Expanded ||
+             index.metric == raft::distance::DistanceType::L2Unexpanded) {
+    detail::rbc_build_index(handle, index, k, detail::EuclideanFunc());
+  } else if (index.metric == raft::distance::DistanceType::L2SqrtExpanded ||
+             index.metric == raft::distance::DistanceType::L2SqrtUnexpanded) {
+    detail::rbc_build_index(handle, index, k, detail::EuclideanFunc());
+
+    // TODO: Call sqrt on distances
+  }
+
+  index.set_index_trained();
+}
+
+/**
+ * Performs a faster exact knn in metric spaces using the triangle
+ * inequality with a number of landmark points to reduce the
+ * number of distance computations from O(n^2) to O(sqrt(n))
+ * @tparam value_idx
+ * @tparam value_t
+ * @tparam value_int
+ * @param handle
+ * @param X
+ * @param m
+ * @param n
+ * @param k
+ * @param inds
+ * @param dists
+ * @param n_samples
+ */
+template <typename value_idx = int64_t, typename value_t,
+          typename value_int = int>
+inline void rbc_all_knn_query(const raft::handle_t &handle,
+                              BallCoverIndex<value_idx, value_t> &index, int k,
+                              value_idx *inds, value_t *dists,
+                              bool perform_post_filtering = true,
+                              float weight = 1.0) {
+  ASSERT(index.n == 2,
+         "Random ball cover currently only works in 2-dimensions");
+  if (index.metric == raft::distance::DistanceType::Haversine) {
+    detail::rbc_all_knn_query(handle, index, k, inds, dists,
+                              detail::HaversineFunc(), perform_post_filtering,
+                              weight);
+  } else if (index.metric == raft::distance::DistanceType::L2Expanded ||
+             index.metric == raft::distance::DistanceType::L2Unexpanded) {
+    detail::rbc_all_knn_query(handle, index, k, inds, dists,
+                              detail::EuclideanFunc(), perform_post_filtering,
+                              weight);
+  } else if (index.metric == raft::distance::DistanceType::L2SqrtExpanded ||
+             index.metric == raft::distance::DistanceType::L2SqrtUnexpanded) {
+    detail::rbc_all_knn_query(handle, index, k, inds, dists,
+                              detail::EuclideanFunc(), perform_post_filtering,
+                              weight);
+
+    // TODO: Call sqrt on distances
+  }
+
+  index.set_index_trained();
+}
+
+template <typename value_idx = int64_t, typename value_t,
```

Comment:
I'd give slight preference to `std::int64_t` and `#include <cstdint>` over `int64_t` and `#include <stdint.h>`. Basically the same argument as for `std::size_t`: using a C type should be an indication to future developers that we are interfacing with C code, and C++ types should otherwise be the default.